### PR TITLE
Only use one direction of numerical grad

### PR DESCRIPTION
**Pull request by unnonouno**
Gradient check requires too long time. I replaced a test for `f(x, y, z)` with a test for `g(a) = f(ax, ay, az)`. Here `g'(a) = f'(ax, ay, az) \dot (x, y, z)`. We can test the implementation of gradient to compare numerical gradient of `g` and dot product of `f'` and `(x, y, z)`.

### CODE REVIEW

**Comment by delta2323** [comment]
If my understanding is correct, it calculates the derivative of `g` at `a=1`, doesn't it?

**Comment by unnonouno** [comment]
Yes. When `a=1`, it checks if `g'(1) = f'(x, y, z) \dot (x, y, z)`.

**Comment by unnonouno** [comment]
Please merge #2978 first.

File: chainer/gradient_check.py [hunk 1]
```diff
@@ -241,40 +241,77 @@ def check_backward(func, x_data, y_grad, params=(),
...
+            data = (one * data).astype(cx.data.dtype)
+            if numpy.isscalar(data):
+                data = xp.array(data)
+            cx.data = data
+        for param, data in six.moves.zip(params, param_data):
+            if dtype is not None:
+                param_dtype = dtype
+            else:
+                param_dtype = param.dtype
+            param.data = (one * data.astype(param_dtype)).astype(param_dtype)
```
Comment (niboshi) [inline chainer/gradient_check.py]: Do we need the inner `.astype`?

File: chainer/gradient_check.py [hunk 2]
```diff
@@ -241,40 +241,77 @@ def check_backward(func, x_data, y_grad, params=(),
...
             assert x.grad is None
-            continue
-        gx, = numerical_grad(f, (cx.data,), y_grad, eps=eps)
-        testing.assert_allclose(gx, x.grad, atol=atol, rtol=rtol)
-        if dtype is None:
-            assert gx.dtype == x.grad.dtype
         else:
-            assert gx.dtype.kind == 'f' and gx.dtype == dtype
+            if x.grad is None:
+                raise TypeError(
```
Comment (niboshi) [inline chainer/gradient_check.py]: How about `RuntimeError`?

**Comment by niboshi** [comment]
Please check comments.
Also, please write some comment about what the code is doing, like what you posted above.

**Comment by niboshi** [comment]
Do you have time for this PR? I want this merged before stabilizing unit tests.

File: chainer/gradient_check.py (hunk 2 above)
Comment (unnonouno) [inline chainer/gradient_check.py]: OK

**Comment by unnonouno** [comment]
I updated it.

**Comment by niboshi** [comment]
Please add some comment as well. It's not apparent.

> Also, please write some comment about what the code is doing, like what you posted above.

**Comment by unnonouno** [comment]
Done!

**Comment by unnonouno** [comment]
Updated

**Comment by niboshi** [comment]
Don't we backport this PR?
IMO test stabilization is also essential for stable branch.

**Comment by unnonouno** [comment]
Humm. Ok, this PR itself does not change behavior of Chainer itself.

File: chainer/gradient_check.py [hunk 3]
```diff
@@ -289,6 +289,9 @@ def g():
         for skip, cx, data in six.moves.zip(no_grads, casted_xs, casted_data):
             if skip:
                 continue
+            # Inner astype is required to execute __mul__ with the given
+            # type, and outer astype is require to store data with the given
+            # type
```
Comment (niboshi) [inline chainer/gradient_check.py]: Could you describe why `__mul__` is required to be executed separately?

Comment (unnonouno) [inline chainer/gradient_check.py]: I found that inner astype is not required.

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**